packages/axios-extension/src/abap/lrep-service.ts

```diff
@@ -0,0 +1,163 @@
+ */
+export interface Message {
+    severity: 'Success' | 'Warning' | 'Error';
+    text: string;
+    details?: string[];
+    id: string;
+    variables?: string[];
+}
+
+/**
+ * Technically supported layers, however, in practice only `CUSTOMER_BASE` is used
+ */
+type Layer = 'VENDOR' | 'CUSTOMER_BASE';
+
+/**
+ * Returns the namespace as string.
+ *
+ * @param namespace either as string or as object
+ * @returns serialized namespace
+ */
+function getNamespaceAsString(namespace: Namespace): string {
+    return namespace['id'] ? `apps/${namespace['reference']}/appVariants/${namespace['id']}/` : (namespace as string);
+}
+
+/**
+ * A class respresenting the design time adaptation service allowing to deploy adaptation projects to an ABAP system.
+ */
+export abstract class LayeredRepositoryService extends Axios implements Service {
+    public static readonly PATH = '/sap/bc/lrep';
+
+    public log: Logger;
+
+    /**
+     * Check whether a variant with the given namespace already exists.
+     *
+     * @param namespace either as string or as object
+     * @returns the Axios response object for futher processing
+     */
+    public async check(namespace: Namespace): Promise<AxiosResponse> {
+        return this.get('/dta_folder/', {
+            params: {
+                name: getNamespaceAsString(namespace),
+                layer: 'CUSTOMER_BASE' as Layer
+            }
+        });
+    }
+
+    /**
+     * Fetch a csrf token required for deployment. The token will be returned as header 'x-csrf-token'.
+     *
+     * @returns the Axios response object for futher processing
+     */
+    public async getCsrfToken(): Promise<AxiosResponse> {
+        return this.get('/actions/getcsrftoken/', {
+            headers: {
+                'X-Csrf-Token': 'Fetch'
+            }
+        });
+    }
+
+    /**
+     * Deploy the given archive either by creating a new folder in the layered repository or updating an existing one.
+     *
+     * @param archivePath path to a zip archive containing the adaptation project
+     * @param config adataption project deployment configuration
+     * @returns the Axios response object for futher processing
+     */
+    public async deploy(archivePath: string, config: AdaptationConfig): Promise<AxiosResponse> {
+        const archive = readFileSync(archivePath);
+
+        const tokenResponse = await this.getCsrfToken();
+        const checkResponse = await this.check(config.namespace);
+        const params: object = {
+            name: getNamespaceAsString(config.namespace),
+            layer: 'CUSTOMER_BASE' as Layer
+        };
+        if (config.package) {
+            params['package'] = config.package;
+            if (config.package.toLowerCase() !== '$tmp') {
+                params['changeList'] = config.transport;
+            }
+        }
+        const response = await this.request({
+            method: checkResponse.status === 200 ? 'PUT' : 'POST',
+            url: '/dta_folder/',
```

Comment:
Use a constant?

Reply:
Could I say no? It is only used twice and I doubt it will ever be changed, so I find the code easier readable with the path as plain text.

Reply:
Of course, code readability trumps the rule of using constants over literals.

Reply:
Well ... after I said it, I added the `undeploy` function and needed it again, and I couldn't get myself to have the same string three times in the file, so now it is a constant.